### PR TITLE
[tf/gcp][fullnode] customize autoscaler and NAT

### PR DESCRIPTION
**Pull request by rustielin**
### Description

Option to make the cluster have public nodes + remove the NAT, such that all GKE worker nodes have a different outbound IP address, and also add the option to enable and configure the autoscaler

### Test Plan

Apply

- - -
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/5043)

### CODE REVIEW

**Review by sherry-x — APPROVED** [review]

File: terraform/fullnode/gcp/variables.tf [hunk 1]
```diff
@@ -148,3 +148,28 @@ variable "enable_kube_state_metrics" {
   description = "Enable kube-state-metrics within monitoring helm chart"
   default     = false
 }
+
+variable "gke_enable_private_nodes" {
+  description = "Enable private nodes for GKE cluster"
+  default     = true
+}
+
+variable "gke_enable_autoscaling" {
```
Comment (perryjrandall) [inline terraform/fullnode/gcp/variables.tf:157]: why not default enable?

**Review by perryjrandall — APPROVED** [review]

File: terraform/fullnode/gcp/variables.tf (hunk 1 above)
Comment (rustielin) [reply terraform/fullnode/gcp/variables.tf:157]: actually, good point

Comment (rustielin) [reply terraform/fullnode/gcp/variables.tf:157]: I think there's a fear of the autoscaler potentially running up the cost, but that depends on operators knowing to change `gke_autoscaling_max_cpu` and `gke_autoscaling_max_memory`

File: terraform/fullnode/gcp/cluster.tf [hunk 2]
```diff
@@ -4,6 +4,13 @@ resource "google_container_cluster" "aptos" {
   location = local.zone
   network  = google_compute_network.aptos.id
 
+  lifecycle {
+    ignore_changes = [
```
Comment (geekflyer) [inline terraform/fullnode/gcp/cluster.tf:8]: this prob shouldn't be in our production module.

File: terraform/fullnode/gcp/cluster.tf [hunk 3]
```diff
@@ -80,4 +103,12 @@ resource "google_container_node_pool" "fullnodes" {
       mode = "GKE_METADATA"
     }
   }
+
+  dynamic "autoscaling" {
+    for_each = var.gke_enable_autoscaling ? [1] : []
```
Comment (geekflyer) [inline terraform/fullnode/gcp/cluster.tf:108]: just fyi, the autoscaling here (on the nodepool) has nothing to do with the autoscaling on the cluster level.

The autoscaling on the nodepool enables autoscaling for this specific nodepool whereas the cluster-level one enables nodepool auto provisioning which automatically provisions new nodepools based on workload requirements which then also have autoscaling enabled.

The nodepool auto provisioning is for most practical purposes superior to enabling single nodepool autoscaling.

Imo the best practice here would be to not create a nodepool explicitly at all if nodepool / cluster auto provisioning is enabled (we're doing this internally for the other non node clusters already btw).

Anyways. not 100% required to change this in this PR, but I just wanted to explain it right away and that we should probably just remove the nodepool here altogether.

File: terraform/fullnode/gcp/variables.tf [hunk 4]
```diff
@@ -148,3 +148,28 @@ variable "enable_kube_state_metrics" {
...
+  description = "Enable autoscaling for the GKE cluster"
```
Comment (geekflyer) [inline terraform/fullnode/gcp/variables.tf:158]: probably better to rename this to `Enable node auto-provisioning` and add a link to it https://cloud.google.com/kubernetes-engine/docs/how-to/node-auto-provisioning.

Unfortunately enable autoscaling for the GKE cluster can mean many things :)

**Review by geekflyer — CHANGES_REQUESTED** [review]
Currently this change conflates essentially 2 seperate features in GKE:
Explicit nodepools with autoscaling and Node Autoprovisioning. Better to make that seperation more clear and use just one or the other. It's actually rarely useful to use them both in the same cluster imo.

**Pull request merged**